Fix: Bug in is_subset function of set handlers.

shashankks0987 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #19513

#### Brief description of what is fixed or changed
earlier, imageset(Lambda(n, 1/n), Integers).is_subset(Reals) incorrectly returned True
now, imageset(Lambda(n, 1/n), Integers).is_subset(Reals) returns None

- Add test for the above scenario

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
  * Fixed a bug in is_subset()
<!-- END RELEASE NOTES -->

shashankks0987 committed:
Fix: Bug in is_subset function of set handlers.
earlier, imageset(Lambda(n, 1/n), Integers).is_subset(Reals) incorrectly returned True
now, imageset(Lambda(n, 1/n), Integers).is_subset(Reals) returns None

- Add test for the above scenario

shashankks0987 committed:
Remove redundant local imports

smichr commented on sympy/sets/handlers/intersection.py:
```diff
@@ -277,9 +277,9 @@ def intersection_sets(self, other): # noqa:F811
                 return FiniteSet(*(fn.subs(n, s[0]) for s in solns))
 
     if other == S.Reals:
-        from sympy.solvers.solveset import solveset_real
```
Although this PR can be patched to use `solve_linear` isn't it better to use the full `solveset_real`? Or maybe you are thinking to keep that for the `simplify` method? But as of now, `simplify` doesn't do anything to the ConditionSet.

In addition, you are incorrectly passing args to `solve_linear`; check parameters.

shashankks0987 replied:
Based on the limited knowledge I have, I thought using solveset_real would be computationally expensive. Although using solveset_real would make this implementation a lot simpler

shashankks0987 replied:
@smichr Should I use solveset_real or is this cool? I've fixed the other stuff

smichr replied:
There is no choice but to use `solveset_real` or else even a simple system's intersection won't be computed. Is there another method that will resolve this simple failing case? `solve_linear` is ok at instantiation where we don't want to work too hard, but for the `intersection` method, we have to do our best to give a solution.

And you are still not using `solve_linear` correctly. Try solving a simple equation like `x + 1 = 0` after reading the docstring and confirm that you get `(x, -1)` as the solution.

oscarbenjamin replied:
A function like `solveset_real` can be very slow in some cases. I don't think that intersection or instantiation should call a potentially slow function like this. It is actually especially bad in something like an intersection handler because these handlers are called for all *pairs* of sets in an intersection. That makes computing an intersection of a large number of sets very slow e.g.:
```python
In [9]: s = [ImageSet(Lambda(n, x**n), Integers) for x in range(2, 100)]

In [10]: Intersection(*s)
^C---------------------------------------------------------------------------
KeyboardInterrupt  
```
If there is a need to call `solveset` then the whole operation should be handled as part of simplification rather than in any part of evaluation. The intersect handlers need to be extremely fast because they are themselves called in an algorithmically bad `O(n**2)` way.

smichr replied:
OK. Then we need to add this to `simplify` at some point. As for the issue here, it is easy to make `solve_linear` still work for the given example by arg-wise calling `solve_linear`:
```diff
diff --git a/sympy/sets/handlers/intersection.py b/sympy/sets/handlers/intersection.py
index 2a34869063..3691b864eb 100644
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -1,5 +1,5 @@
 from sympy import (S, Dummy, Lambda, symbols, Interval, Intersection, Set,
-                   EmptySet, FiniteSet, Union, ComplexRegion)
+                   EmptySet, FiniteSet, Union, ComplexRegion, Mul)
 from sympy.multipledispatch import dispatch
 from sympy.sets.conditionset import ConditionSet
 from sympy.sets.fancysets import (Integers, Naturals, Reals, Range,
@@ -304,16 +304,16 @@ def intersection_sets(self, other): # noqa:F811
             return None
         else:
             # univarite imaginary part in same variable
-            x, xis = solve_linear(im, n)
-            if x == n:
+            x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
+            if x and all(i == n for i in x):
                 base_set -= FiniteSet(xis)
             else:
                 base_set -= ConditionSet(n, Eq(im, 0), S.Integers)
         # exclude values that make denominators 0
         for i in denoms(f):
             if i.has(n):
-                x, xis = solve_linear(i, n)
-                if x == n:
+                x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
+                if x and all(i == n for i in x):
                     base_set -= FiniteSet(xis)
                 else:
                     base_set -= ConditionSet(n, Eq(i, 0), S.Integers)
```

shashankks0987 committed:
Fix: Incorrectly passed paramters to solveset_linear in intersection.py
- Fix failing test in test_fancysets

smichr commented on sympy/sets/handlers/intersection.py:
```diff
@@ -304,7 +304,7 @@ def intersection_sets(self, other): # noqa:F811
         else:
             # univarite imaginary part in same variable
-            x, xis = solve_linear(im, n)
+            x, xis = solve_linear(n, im)
```
this solves the expression `n = im`...and you need to solve `im == 0`.

shashankks0987 replied:
ok understood

shashankks0987 committed:
Fix: Usage of solve_linear
- Handle cases where solve_linear returns an empty list
- Fix failing tests
- Remove redundant imports

shashankks0987 commented on sympy/sets/handlers/intersection.py:
```diff
@@ -304,17 +304,19 @@ def intersection_sets(self, other): # noqa:F811
-                if x == n:
-                    base_set -= FiniteSet(xis)
+                sol = list(zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols]))
+                if sol != []:
```
This is because solve_linear sometimes returns an empty list in some cases

oscarbenjamin commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -656,9 +657,7 @@ def test_infinitely_indexed_set_2():
-    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == \
-            FiniteSet(-1, 1)
-
+    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == Complement(S.Integers, FiniteSet((-1, 1)))
```
This doesn't seem correct

oscarbenjamin replied:
@smichr @shashankks0987 this output is incorrect:
```python
In [8]: S1 = imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers)

In [9]: S1
Out[9]: {n + ⅈ⋅(n - 1)⋅(n + 1) │ n ∊ ℤ}

In [10]: 2 in S1
Out[10]: False

In [11]: 2 in S1.intersect(Reals)
Out[11]: True
```
What is actually being returned is:
```python
In [12]: S1.intersect(S2)
Out[12]: ℤ \ {(-1, 1)}
```
That's the set of integers minus the set containing a single tuple `(-1, 1)`. Something has clearly gone wrong there. The previous answer was correct. I think it's fine if this intersection doesn't evaluate but it shouldn't return a wrong result like this.

shashankks0987 replied:
Ok I'll try fixing this

shashankks0987 replied:
```
In [4]: S1
Out[4]: {n + ⅈ⋅(n - 1)⋅(n + 1) │ n ∊ ℤ}

In [5]: 2 in S1
Out[5]: False

In [6]: 2 in S1.intersect(Reals)
Out[6]: False

In [7]: S2 = Reals

In [8]: S1.intersect(S2)
Out[8]: {-1, 1}
```
This seems to be working fine though

shashankks0987 replied:
With the new diff @oscarbenjamin

shashankks0987 commented:
@oscarbenjamin Does this work? (I'm assuming other cases are tested)
```diff
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -306,7 +306,7 @@ def intersection_sets(self, other): # noqa:F811
             # univarite imaginary part in same variable
             x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
             if x and all(i == n for i in x):
-                base_set -= FiniteSet(xis)
+                base_set = FiniteSet(*xis)
             else:
                 base_set -= ConditionSet(n, Eq(im, 0), S.Integers)
         # exclude values that make denominators 0
diff --git a/sympy/sets/tests/test_fancysets.py b/sympy/sets/tests/test_fancysets.py
index 098020ca4e..dd67662f01 100644
--- a/sympy/sets/tests/test_fancysets.py
+++ b/sympy/sets/tests/test_fancysets.py
@@ -657,7 +657,7 @@ def test_infinitely_indexed_set_2():
 def test_imageset_intersect_real():
     from sympy import I
     from sympy.abc import n
-    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == Complement(S.Integers, FiniteSet((-1, 1)))
+    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == FiniteSet(-1, 1)
```

oscarbenjamin commented:
> I'm really not sure this is right though, I feel it's a hack ><

Why does it seem like a hack?

The code as merged definitely has a bug and what you've shown at least fixes it for the given test case.

shashankks0987 commented:
> > I'm really not sure this is right though, I feel it's a hack ><
> 
> Why does it seem like a hack?
> 
> The code as merged definitely has a bug and what you've shown at least fixes it for the given test case.

I feel so because we're assigning the unpacked values to the solution set rather than subtracting/adding from it,
Anyways I'll raise a PR for now as it is still better than the merged code

oscarbenjamin commented:
Yes, that doesn't seem write. I think it should be `FiniteSet(*xis)` in any case but maybe something else is wrong.
